File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
from collections import deque

input = sys.stdin.readline

N, M = map(int, input().split())

visited = [False] * (N + 1)
graph = [[0] * (N + 1) for i in range(N+1)]

for _ in range(M):
    a, b = map(int, input().split())
    graph[a][b] = graph[b][a] = 1

count = 0

def dfs(v):
    visited[v] = True
    for i in range(1, N+1):
        if visited[i] == False and graph[v][i] == 1:
            dfs(i)


for k in range(1, N+1):
    if visited[k] == False:
        count += 1
        dfs(k)


print(count)

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : fs.readFileSync("../input.txt").toString()
).split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const [N, M] = input()
  .trim()
  .split(" ")
  .map((v) => parseInt(v));

const visited = new Array(N + 1).fill(false);
const arr = [];
for (let i = 0; i < M; i++) {
  arr.push(
    input()
      .trim()
      .split(" ")
      .map((v) => parseInt(v))
  );
}

let count = 0;

const graph = {};

for (let i = 1; i < N + 1; i++) {
  graph[i] = [];
}

arr.forEach((v) => {
  graph[v[0]].push(v[1]);
  graph[v[1]].push(v[0]);
});

function dfs(index) {
  if (visited[index] || visited[index].length === 0) return;
  visited[index] = true;
  graph[index].forEach((v) => {
    dfs(v);
  });
  return true;
}

for (let i = 1; i < N + 1; i++) {
  if (visited[i] === false) {
    const response = dfs(i);
    if (response) count++;
  }
}

console.log(count);